feature/experimental/gribi/ate_tests/route_addition_during_failover_test/route_addition_during_failover_test.go

```diff
@@ -0,0 +1,668 @@
+
+	t.Log("Flush existing gRIBI routes before test.")
+	if err := gribi.FlushAll(client); err != nil {
+		t.Fatal(err)
+	}
+
+	args := &testArgs{
+		ctx:        ctx,
+		client:     client,
+		dut:        dut,
+		ate:        ate,
+		top:        top,
+		electionID: eID,
+	}
+
+	virtualIPs := createIPv4Entries(t, ipBlock1FlowArgs.ipBlock)
+
+	t.Log("Inject routes from ipBlock1 in default VRF with NHGID: #1.")
+	pushDefaultEntries(t, args, subIntfIPs, virtualIPs, configNhg, !switchover)
+
+	gr, err := args.client.Get().
+		WithNetworkInstance(*deviations.DefaultNetworkInstance).
+		WithAFT(fluent.IPv4).
+		Send()
+	if err != nil {
+		t.Fatalf("Error encountered during gRIBI Get operation: got: %v", err)
+	}
+
+	for _, r := range gr.GetEntry() {
+		entry := r.GetIpv4().Prefix
+		if got := r.GetFibStatus().String(); got != "PROGRAMMED" {
+			t.Fatalf("gRIBI entry %v is not programmed in FIB, got:%v, want: PROGRAMMED", entry, got)
+		}
+	}
+
+	// Send traffic from ATE port-1 to prefixes in ipBlock1 and ensure traffic
+	// flows 100% and reaches ATE port-2.
+	testTraffic(t, *args, createTrafficFlow(t, args.ate, args.top, ipBlock1FlowArgs))
+
+	controllers := cmp.FindComponentsByType(t, dut, controllerCardType)
+	t.Logf("Found controller list: %v", controllers)
+	// Only perform the switchover for the chassis with dual controllers.
+	if len(controllers) != 2 {
+		t.Skipf("Dual controllers required on %v: got %v, want 2", dut.Model(), len(controllers))
+	}
+
+	secondaryBeforeSwitch, primaryBeforeSwitch := findController(t, dut, controllers)
+
+	awaitSwitchoverReady(t, dut, primaryBeforeSwitch)
+	t.Logf("Controller %q is ready for switchover before test.", primaryBeforeSwitch)
+
+	gnoiClient := dut.RawAPIs().GNOI().Default(t)
+	switchoverRequest := &spb.SwitchControlProcessorRequest{
+		ControlProcessor: cmp.GetSubcomponentPath(secondaryBeforeSwitch),
+	}
+	t.Logf("switchoverRequest: %v", switchoverRequest)
+
+	// Concurrently run switchover and gribi route addition in ipBlock2.
+	virtualIPsBlock2 := createIPv4Entries(t, ipBlock2FlowArgs.ipBlock)
+
+	// Check for coredumps in the DUT and validate that none are present on DUT before switchover.
+	coreFileCheck(t, dut, gnoiClient, sysConfigTime)
+
+	t.Log("Execute gRIBi route addition and master switchover concurrently.")
+	go func(msg string) {
+		t.Log("Inject routes from ipBlock2 in default VRF with NHGID: #1.")
+		pushDefaultEntries(t, args, subIntfIPs, virtualIPsBlock2, !configNhg, switchover)
+	}("gRIBi Add")
+
+	go func(msg string) {
+		switchoverResponse, err := gnoiClient.System().SwitchControlProcessor(context.Background(), switchoverRequest)
+		if err != nil {
+			t.Logf("Failed to perform control processor switchover with unexpected err: %v", err)
+		}
+		t.Logf("gnoiClient.System().SwitchControlProcessor() response: %v, err: %v", switchoverResponse, err)
+	}("Master Switchover")
+
+	t.Log("Concurrent switchover and route addition is completed, validate switchoverStatus now.")
```

Comment:
how do you know they're complete by this point?

Reply:
@greg-dennis , as per testplan we are triggering concurrent gribi add followed by switchover.  
During swithcover we might not receive response for gribi add route and we are validating switchover state using validateSwitchoverStatus.

Should we add Wait here and check if go concurrent functions are completed?

Reply:
To make sure the concurrent functions are done by this point, you can use a WaitGroup like this:

```
wg := new(sync.WaitGroup)
wg.Add(2)
go func() {
	defer wg.Done()
	t.Log("Inject routes from ipBlock2 in default VRF with NHGID: #1.")
	pushDefaultEntries(t, args, subIntfIPs, virtualIPsBlock2, !configNhg, switchover)
}()

go func() {
	defer wg.Done()
	switchoverResponse, err := gnoiClient.System().SwitchControlProcessor(context.Background(), switchoverRequest)
	if err != nil {
		t.Logf("Failed to perform control processor switchover with unexpected err: %v", err)
	}
	t.Logf("gnoiClient.System().SwitchControlProcessor() response: %v, err: %v", switchoverResponse, err)
}()
wg.Wait()
t.Log("Concurrent switchover and route addition is completed, validate switchoverStatus now.")
```

Reply:
updated waitgroups, thank you.